signer/tests/integration/transaction_coordinator.rs

```diff
@@ -144,9 +154,316 @@ where
+                Box::pin(async {
+                    Ok(ContractSrcResponse {
+                        source: "".to_string(),
+                        publish_height: 1,
+                        marf_proof: None,
+                    })
+                })
+            });
+
+        // For the following 4 contracts we will return 404.
+        // So the coordinator will try to deploy them.
+        client
+            .expect_get_contract_source()
+            .times(4)
+            .returning(|_, _| {
+                Box::pin(async {
+                    Err(Error::StacksNodeResponse(
+                        mock_reqwests_status_code_error(404).await,
+                    ))
+                })
+            });
+
+        // All the following functions, `estimate_fees`, `get_account` and `submit_tx`
+        // are only called 5 times for the contracts to be deployed and 1 for the deposit tx
+        client
+            .expect_estimate_fees()
+            .times(5)
+            .returning(|_, _| Box::pin(async { Ok(100) }));
+
+        // `get_account` will be called 7 times, 2 for the first try to deploy the contracts,
+        // 4 for the second try and 1 for the deposit tx
+        client.expect_get_account().times(7).returning(move |_| {
+            Box::pin(async move {
+                Ok(AccountInfo {
+                    balance: 1_000_000,
+                    locked: 0,
+                    unlock_height: 0,
+                    nonce: nonce,
+                })
+            })
+        });
+        // `submit_tx` will be called 6 times for the contracts to be deployed and 1 for the deposit tx
+        client.expect_submit_tx().times(6).returning(move |tx| {
+            let tx = tx.clone();
+            let txid = tx.txid();
+            let broadcasted_transaction_tx = broadcasted_transaction_tx.clone();
+            Box::pin(async move {
+                broadcasted_transaction_tx
+                    .send(tx)
+                    .expect("Failed to send result");
+                Ok(SubmitTxResponse::Acceptance(txid))
+            })
+        });
+    })
+}
+
+#[ignore = "These tests take ~10 seconds per test case to run"]
+#[test_case(6, &CONTRACT_DEPLOY_INFO, mock_deploy_all_contracts; "deploy-all-contracts")]
+#[test_case(4, &CONTRACT_DEPLOY_INFO[2..], mock_deploy_remaining_contracts_when_some_already_deployed; "deploy-remaining-contracts-when-some-already-deployed")]
+#[test_case(1, &[], mock_deploy_none_when_already_deployed; "deploy-none-when-already-deployed")]
+#[test_case(6, &CONTRACT_DEPLOY_INFO, mock_recover_and_deploy_all_contracts_after_failure; "recover-and-deploy-all-contracts-after-failure")]
 #[tokio::test]
-async fn process_complete_deposit() {
+async fn process_complete_deposit(
+    num_messages: usize,
+    deployed_contracts_data: &[(&str, &str)],
+    stacks_client_mock: impl FnOnce(
+        u64,
+        Sender<StacksTransaction>,
+    ) -> Box<dyn FnOnce(&mut MockStacksInteract)>,
```

Comment:
I plan on doing a little more work here to break out this test into one that tests complete deposit and another that tests contract deployment.

Reply:
I've updated this